File: resources/expected/ros2/building_control_gen_mixed_inverted_topics/strict/src/building_control_cpp_pkg/src/base_code/tcp_fan_base_src.cpp
#include "building_control_cpp_pkg/base_headers/tcp_fan_base_src.hpp"

//========================================================
// Re-running Codegen will overwrite changes to this file
//========================================================

tcp_fan_base::tcp_fan_base() : Node("tcp_fan")
{
    cb_group_ = this->create_callback_group(rclcpp::CallbackGroupType::Reentrant);
    subscription_options_.callback_group = cb_group_;

    // Setting up connections
    tcp_fan_fanCmd_subscription_ = this->create_subscription<building_control_cpp_pkg_interfaces::msg::FanCmd>(
        "tcp_tempControl_fanCmd",
        1,
        std::bind(&tcp_fan_base::accept_fanCmd, this, std::placeholders::_1), subscription_options_);

    tcp_fan_fanAck_publisher_ = this->create_publisher<building_control_cpp_pkg_interfaces::msg::FanAck>(
        "tcp_fan_fanAck",
        1);

    // Used by receiveInputs
    inDataPortTupleVector = {
    };

    // Used by sendOutputs
    outPortTupleVector = {
        {&applicationOut_fanAck, &infrastructureOut_fanAck, &tcp_fan_base::sendOut_fanAck}
    };

}

//=================================================
//  C o m m u n i c a t i o n
//=================================================

void tcp_fan_base::accept_fanCmd(building_control_cpp_pkg_interfaces::msg::FanCmd msg)
{
    enqueue(infrastructureIn_fanCmd, msg);
    std::thread([this]() {
        std::lock_guard<std::mutex> lock(mutex_);
        receiveInputs(infrastructureIn_fanCmd, applicationIn_fanCmd);
        if (applicationIn_fanCmd.empty()) return;
        handle_fanCmd_base(applicationIn_fanCmd.front());
        applicationIn_fanCmd.pop();
        sendOutputs();
    }).detach();
}

void tcp_fan_base::handle_fanCmd_base(MsgType msg)
{
    if (auto typedMsg = std::get_if<building_control_cpp_pkg_interfaces::msg::FanCmd>(&msg)) {
        handle_fanCmd(*typedMsg);
    } else {
        PRINT_ERROR("Receiving wrong type of variable on port fanCmd.\nThis shouldn't be possible.  If you are seeing this message, please notify this tool's current maintainer.");
    }
}

void tcp_fan_base::sendOut_fanAck(MsgType msg)
{
    if (auto typedMsg = std::get_if<building_control_cpp_pkg_interfaces::msg::FanAck>(&msg)) {
        tcp_fan_fanAck_publisher_->publish(*typedMsg);
    } else {
        PRINT_ERROR("Sending out wrong type of variable on port fanAck.\nThis shouldn't be possible.  If you are seeing this message, please notify this tool's current maintainer.");
    }
}

void tcp_fan_base::put_fanAck(building_control_cpp_pkg_interfaces::msg::FanAck msg)
{
    enqueue(applicationOut_fanAck, msg);
}

void tcp_fan_base::receiveInputs(std::queue<MsgType>& infrastructureQueue, std::queue<MsgType>& applicationQueue) {
    if (!infrastructureQueue.empty()) {
        MsgType eventMsg = infrastructureQueue.front();
        infrastructureQueue.pop();
        enqueue(applicationQueue, eventMsg);
    }

    for (std::tuple<std::queue<MsgType>*, std::queue<MsgType>*> port : inDataPortTupleVector) {
        auto infrastructureQueue = std::get<0>(port);
        if (!infrastructureQueue->empty()) {
            auto msg = infrastructureQueue->front();
            enqueue(*std::get<1>(port), msg);
        }
    }
}

void tcp_fan_base::enqueue(std::queue<MsgType>& queue, MsgType val) {
    if (queue.size() >= 1) {
        queue.pop();
    }
    queue.push(val);
}

void tcp_fan_base::sendOutputs() {
    for (std::tuple<std::queue<MsgType>*, std::queue<MsgType>*, void (tcp_fan_base::*)(MsgType)> port : outPortTupleVector) {
        auto applicationQueue = std::get<0>(port);
        if (applicationQueue->size() != 0) {
            auto msg = applicationQueue->front();
            applicationQueue->pop();
            enqueue(*std::get<1>(port), msg);
        }
    }

    for (std::tuple<std::queue<MsgType>*, std::queue<MsgType>*, void (tcp_fan_base::*)(MsgType)> port : outPortTupleVector) {
        auto infrastructureQueue = std::get<1>(port);
        if (infrastructureQueue->size() != 0) {
            auto msg = infrastructureQueue->front();
            infrastructureQueue->pop();
            (this->*std::get<2>(port))(msg);
        }
    }
}

File: resources/expected/ros2/python-building_control_gen_mixed_inverted_topics/strict/src/building_control_py_pkg/building_control_py_pkg/base_code/tcp_fan_base_src.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from collections import deque
from typing import Union
import threading
from rclpy.callback_groups import ReentrantCallbackGroup
from building_control_py_pkg_interfaces.msg import FanCmd
from building_control_py_pkg_interfaces.msg import FanAck

#========================================================
# Re-running Codegen will overwrite changes to this file
#========================================================

class tcp_fan_base(Node):
    def __init__(self):
        super().__init__("tcp_fan")

        self.cb_group_ = ReentrantCallbackGroup()

        self.lock_ = threading.Lock()

        # Setting up connections
        self.tcp_fan_fanCmd_subscription_ = self.create_subscription(
            FanCmd,
            "tcp_tempControl_fanCmd",
            self.accept_fanCmd,
            1,
            callback_group=self.cb_group_)

        self.tcp_fan_fanAck_publisher_ = self.create_publisher(
            FanAck,
            "tcp_fan_fanAck",
            1)

        self.infrastructureIn_fanCmd = deque()
        self.applicationIn_fanCmd = deque()

        self.infrastructureOut_fanAck = deque()
        self.applicationOut_fanAck = deque()

        # Used by receiveInputs
        self.inDataPortTupleVector = [
        ]

        # Used by sendOutputs
        self.outPortTupleVector = [
            [self.applicationOut_fanAck, self.infrastructureOut_fanAck, self.sendOut_fanAck]
        ]

    def timeTriggered(self):
        raise NotImplementedError("Subclasses must implement this method")

    #=================================================
    #  C o m m u n i c a t i o n
    #=================================================

    def fanCmd_thread(self):
        with self.lock_:
            self.receiveInputs(self.infrastructureIn_fanCmd, self.applicationIn_fanCmd)
            if len(self.applicationIn_fanCmd) == 0: return
            self.handle_fanCmd_base(self.applicationIn_fanCmd[0])
            self.applicationIn_fanCmd.pop()
            self.sendOutputs()

    def accept_fanCmd(self, msg):
        self.enqueue(self.infrastructureIn_fanCmd, msg)
        thread = threading.Thread(target=self.fanCmd_thread)
        thread.daemon = True
        thread.start()


    def sendOut_fanAck(self, msg):
        if type(msg) is FanAck:
            self.tcp_fan_fanAck_publisher_.publish(msg)
        else:
            self.get_logger().error("Sending out wrong type of variable on port fanAck.\nThis shouldn't be possible.  If you are seeing this message, please notify this tool's current maintainer.")

    def put_fanAck(self, msg):
        self.enqueue(self.applicationOut_fanAck, msg)

    #=================================================
    #  C o m p u t e    E n t r y    P o i n t
    #=================================================
    def handle_fanCmd(self, msg):
        raise NotImplementedError("Subclasses must implement this method")

    def handle_fanCmd_base(self, msg):
        if type(msg) is FanCmd:
            self.handle_fanCmd(msg)
        else:
            self.get_logger.error("Receiving wrong type of variable on port fanCmd.\nThis shouldn't be possible.  If you are seeing this message, please notify this tool's current maintainer.")


    def receiveInputs(self, infrastructureQueue, applicationQueue):
        if not(len(infrastructureQueue) == 0):
            eventMsg = infrastructureQueue[0]
            infrastructureQueue.pop()
            self.enqueue(applicationQueue, eventMsg)

        for port in self.inDataPortTupleVector:
            infrastructureQueue = port[0]
            if not(len(infrastructureQueue) == 0):
                msg = infrastructureQueue[0]
                self.enqueue(port[1], msg)

    def enqueue(self, queue, val):
        if len(queue) >= 1:
            queue.pop()
        queue.append(val)

    def sendOutputs(self):
        for port in self.outPortTupleVector:
            applicationQueue = port[0]
            if len(applicationQueue) != 0:
                msg = applicationQueue[0]
                applicationQueue.pop()
                self.enqueue(port[1], msg)

        for port in self.outPortTupleVector:
            infrastructureQueue = port[1]
            if len(infrastructureQueue) != 0:
                msg = infrastructureQueue[0]
                infrastructureQueue.pop()
                (port[2])(msg)
